cpp/pagerank_module/algorithm/pagerank.cpp

```diff
@@ -0,0 +1,147 @@
+                             std::vector<std::uint64_t> &borders) {
+  if (number_of_threads == 0) {
+    throw std::runtime_error("Number of threads can't be zero (0)!");
+  }
+
+  for (std::uint64_t border_index = 0; border_index <= number_of_threads; border_index++) {
+    borders.push_back(border_index * graph.GetEdgeCount() / number_of_threads);
+  }
+}
+
+void ThreadPageRankIteration(const PageRankGraph &graph, const std::vector<double> &old_rank, const std::uint64_t lo,
+                             const std::uint64_t hi, std::promise<std::vector<double>> new_rank_promise) {
+  std::vector<double> new_rank(graph.GetNodeCount(), 0);
+  // Calculate sums of PR(page)/C(page) scores for the entire block (from lo to
+  // hi edges).
+  for (std::size_t edge_id = lo; edge_id < hi; edge_id++) {
+    const auto [source, target] = graph.GetOrderedEdges()[edge_id];
+    // Add the score of target node to the sum.
+    new_rank[source] += old_rank[target] / graph.GetOutDegree(target);
+  }
+  new_rank_promise.set_value(new_rank);
+}
+
+void AddCurrentBlockToRankNext(const PageRankGraph &graph, const double damping_factor,
+                               const std::vector<double> &block, std::vector<double> &rank_next) {
+  // The block vector contains partially precalculated sums of PR(page)/C(page)
+  // for each node. Node index in the block vector corresponds to the node index
+  // in the rank_next vector.
+  for (std::size_t node_index = 0; node_index < block.size(); node_index++) {
+    rank_next[node_index] += damping_factor * block[node_index];
+  }
+}
+
+void CompleteRankNext(const PageRankGraph &graph, const double damping_factor, std::vector<double> &rank_next) {
+  while (rank_next.size() < graph.GetNodeCount()) {
+    rank_next.push_back((1.0 - damping_factor) / graph.GetNodeCount());
+  }
+}
+
+bool CheckContinueIterate(const std::vector<double> &rank, const std::vector<double> &rank_next,
+                          const std::size_t max_iterations, const double stop_epsilon,
+                          const std::size_t number_of_iterations) {
+  if (number_of_iterations == max_iterations) {
+    return false;
+  }
+  for (std::size_t node_id = 0; node_id < rank.size(); node_id++) {
+    if (std::abs(rank[node_id] - rank_next[node_id]) > stop_epsilon) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void NormalizeRank(std::vector<double> &rank) {
+  const double sum = std::accumulate(rank.begin(), rank.end(), 0.0);
+  for (double &value : rank) {
+    value /= sum;
+  }
+}
+
+std::vector<double> ParallelIterativePageRank(const PageRankGraph &graph, std::size_t max_iterations,
+                                              double damping_factor, double stop_epsilon) {
+  const std::uint64_t number_of_threads = std::thread::hardware_concurrency();
+
+  std::vector<std::uint64_t> borders;
+  CalculateOptimalBorders(graph, number_of_threads, borders);
```

Comment:
Just transform this function into one that returns `borders` vector (no need to `return std::move()` because of RVO).